src/main/java/com/yahoo/bullet/querying/QueryManager.java

```diff
@@ -0,0 +1,252 @@
+
+    private static class NoPartitioner implements Partitioner {
+        private static final Set<String> EMPTY_KEYS = Collections.singleton("");
+
+        @Override
+        public Set<String> getKeys(Query query) {
+            return EMPTY_KEYS;
+        }
+
+        @Override
+        public Set<String> getKeys(BulletRecord record) {
+            return EMPTY_KEYS;
+        }
+    }
+
+    /**
+     * The constructor that takes a non-null {@link BulletConfig} instance that contains partitioning settings.
+     *
+     * @param config The non-null config.
+     */
+    public QueryManager(BulletConfig config) {
+        boolean enable = config.getAs(BulletConfig.QUERY_PARTITIONER_ENABLE, Boolean.class);
+        if (enable) {
+            partitioner = config.loadConfiguredClass(BulletConfig.QUERY_PARTITIONER_CLASS_NAME);
+            log.info("Partitioning for queries is enabled. Using %s", partitioner.getClass().getName());
+        } else {
+            partitioner = new NoPartitioner();
+        }
+        partitioning = new HashMap<>();
+        queries = new HashMap<>();
+    }
+
+    /**
+     * Adds a configured, initialized query instance {@link Querier} to the manager.
+     *
+     * @param id The query ID.
+     * @param querier A fully initialized (using {@link Querier#initialize()} {@link Querier} instance.
+     */
+    public void addQuery(String id, Querier querier) {
+        Query query = querier.getQuery();
+        Set<String> keys = partitioner.getKeys(query);
+        for (String key : keys) {
+            Set<String> partition = partitioning.getOrDefault(key, new HashSet<>());
+            partition.add(id);
+            partitioning.put(key, partition);
+            log.debug("Added query: %s to partition: %s", id, key);
+        }
+        queries.put(id, querier);
+    }
+
+    /**
+     * Removes and returns a {@link Querier} from the manager. The manager does not have any information pertaining to
+     * the query any longer.
+     *
+     * @param id The query ID to remove.
+     * @return The removed {@link Querier} instance.
+     */
+    public Querier removeAndGetQuery(String id) {
+        Querier querier = queries.remove(id);
+        if (querier != null) {
+            Query query = querier.getQuery();
+            Set<String> keys = partitioner.getKeys(query);
+            for (String key : keys) {
+                partitioning.get(key).remove(id);
+                log.debug("Removed query: %s from partition: %s", id, key);
+            }
+        }
+        return querier;
+    }
+
+    /**
+     * Removes and returns the {@link List} of {@link Querier} instances for the given non-null query IDs. The manager
+     * does not have any information pertaining to these queries after.
+     *
+     * @param ids The non-null {@link Set} of query IDs to remove.
+     * @return The removed {@link List} of {@link Querier} instances.
+     */
+    public List<Querier> removeAndGetQueries(Set<String> ids) {
```

Comment:
Should this return a set? Since the input param was a set (no concept of order)?

Reply:
We take a Set of query IDs not Queriers but this is a good point. I'd rather leave it otherwise I'd have to implement hashCode and equals for Querier.